### PR TITLE
Fix queue bugs, harden edge cases, add test harness

### PR DESCRIPTION
**Pull request by promiseofcake**
## Summary

- **Fix stale pipeline files** — cleans up leftover `pipeline-*.json` between loop iterations, preventing indefinite queue blocking
- **Accept 2xx HTTP status codes** — `fetch()` now handles 202 Accepted (and other 2xx), fixing cancel workflow requests
- **Fix jq quoting** — adds `-r` flag to all jq string extractions, fixing malformed cancel URLs and fragile timestamp comparisons
- **Same-timestamp tiebreaker** — uses workflow ID as deterministic tiebreaker when two workflows share a creation timestamp
- **URL-encode branch names** — handles special characters in branch names via `jq @uri`
- **Pipeline queue timeout** — adds optional `time` and `dont_quit` parameters, matching global queue (default: wait indefinitely, backwards compatible)
- **Code hygiene** — `local` variables in `fetch()`, debug message typos fixed
- **Test harness** — 15 automated tests with curl mocking, assertion helpers, and fixture data (no real API calls)

## Test plan

- [x] `make test` — 15/15 tests pass (11 global queue, 4 pipeline queue)
- [x] `shellcheck` clean on all scripts
- [ ] Integration test via CircleCI orb pipeline

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by promiseofcake** [comment]
Your development orb(s) have been published. It will expire in 90 days.
You can preview what this will look like on the CircleCI Orb Registry at the following link(s): 
https://circleci.com/developer/orbs/orb/promiseofcake/workflow-queue?version=dev:2548502ecd69e1b5003bf01b8df467b7fce8218c
https://circleci.com/developer/orbs/orb/promiseofcake/workflow-queue?version=dev:alpha

**Comment by promiseofcake** [comment]
Your development orb(s) have been published. It will expire in 90 days.
You can preview what this will look like on the CircleCI Orb Registry at the following link(s): 
https://circleci.com/developer/orbs/orb/promiseofcake/workflow-queue?version=dev:96360de4217b8e619048cb031946e6a72bf8dab9
https://circleci.com/developer/orbs/orb/promiseofcake/workflow-queue?version=dev:alpha

File: docs/plans/2026-02-27-bug-fixes-and-hardening.md [hunk 1]
```diff
@@ -0,0 +1,841 @@
...
+### Task 8: Add test harness with curl mocking
+
+**Problem:** The current test.sh makes real API calls and the workflow-fixture.json is unused. There are no automated assertions.
+
+**Files:**
+- Create: `test/test_global_queue.sh`
+- Create: `test/test_pipeline_queue.sh`
+- Create: `test/helpers.sh` (shared test utilities and mock setup)
+- Create: `test/fixtures/pipelines.json`
+- Create: `test/fixtures/pipeline-workflow-old.json`
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply docs/plans/2026-02-27-bug-fixes-and-hardening.md:384]: Acknowledged — cosmetic mismatch in the plan doc. Not worth a follow-up commit since the plan is already executed and the fixture files are correct.

File: test/test_global_queue.sh [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    source_global_functions
+
+    mock_curl_response() {
+        echo '{"status":"cancelled"}' > "$3"
+        echo "202"
+    }
+    export -f mock_curl_response
+
+    fetch "https://example.com/cancel" "${TMP_DIR}/cancel.json" "POST"
+    assert_equals "fetch returns success for 202" "0" "$?"
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply test/test_global_queue.sh]: Good catch. Fixed in 96360de — now runs fetch in a subshell and captures exit code explicitly, consistent with `test_fetch_fails_on_non_transient_error`.

File: test/test_pipeline_queue.sh [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        {"id": "wf-current-1111", "name": "wf-a", "status": "running", "created_at": "2026-01-15T10:00:00Z"}
+    ]
+}
+FIXTURE
+        echo "200"
+    }
+    export -f mock_curl_response
+
+    workflows_file="${TMP_DIR}/workflow_status.json"
+    fetch_pipeline_workflows > /dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply test/test_pipeline_queue.sh]: Good catch. Removed in 96360de.

File: test/test_pipeline_queue.sh [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        {"id": "wf-other-4444", "name": "wf-c", "status": "created", "created_at": "2026-01-15T10:00:00Z"}
+    ]
+}
+FIXTURE
+        echo "200"
+    }
+    export -f mock_curl_response
+
+    workflows_file="${TMP_DIR}/workflow_status.json"
+    fetch_pipeline_workflows > /dev/null 2>&1
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply test/test_pipeline_queue.sh]: Fixed alongside #3 in 96360de.

File: src/scripts/pipeline_queue.sh [hunk 5]
```diff
@@ -109,5 +118,16 @@ while true; do
...
+    if [ "$max_time_seconds" -gt 0 ] && [ $wait_time -ge $max_time_seconds ]; then
+        echo "Max wait time exceeded, considering response."
+        if [ "${CONFIG_DONT_QUIT}" == "1" ]; then
+            echo "Orb parameter dont-quit is set to true, letting this job proceed!"
+            exit 0
+        else
+            echo "Max wait time exceeded. Failing job."
+            exit 1
+        fi
+    fi
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply src/scripts/pipeline_queue.sh:130]: Acknowledged. Testing the main execution loop as a subprocess would require significant test infrastructure (mocking sleep, controlling loop iterations, handling set -e in subprocesses). The timeout logic is straightforward arithmetic and mirrors the already-proven global queue pattern. The existing tests cover the building blocks (fetch, workflow filtering, counting). Not worth the complexity for now.

File: test/test_global_queue.sh [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+        else
+            echo '{"ok":true}' > "$3"
+            echo "200"
+        fi
+    }
+    export -f mock_curl_response
+
+    # Override sleep to not actually wait
+    sleep() { :; }
+    export -f sleep
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply test/test_global_queue.sh:59]: Fixed in 2e24e83 — `unset -f sleep` after the test to restore the real sleep and prevent leaking into subsequent tests.

File: test/fixtures/pipelines.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+    "next_page_token": null,
+    "items": [
+        { "id": "pipeline-1111" },
+        { "id": "pipeline-2222" }
+    ]
```
[the remark this answers is not in the document]
Comment (promiseofcake) [reply test/fixtures/pipelines.json]: Agreed — deleted the entire `test/fixtures/` directory in 2e24e83. All test data is inline and the fixtures were dead weight.

**Comment by promiseofcake** [comment]
Your development orb(s) have been published. It will expire in 90 days.
You can preview what this will look like on the CircleCI Orb Registry at the following link(s): 
https://circleci.com/developer/orbs/orb/promiseofcake/workflow-queue?version=dev:2e24e8391177931797a5f171b3b0abd16a0c4945
https://circleci.com/developer/orbs/orb/promiseofcake/workflow-queue?version=dev:alpha
